pkg/sql/opt_index_selection.go

```diff
@@ -1153,15 +1167,21 @@ func spanFromLogicalSpan(
 		if err != nil {
 			return roachpb.Span{}, err
 		}
+
 		if !part.inclusive {
 			if i != len(ls.end)-1 {
 				return roachpb.Span{}, errors.New("exclusive end constraint must be last")
 			}
-			return s, nil
+
+			lastPartInclusive = false
 		}
 	}
-	s.EndKey = s.EndKey.PrefixEnd()
-	return s, nil
+
+	// We tighten the end key to prevent reading interleaved children
+	// after the last parent key.
+	s.EndKey, err = sqlbase.TightenEndKey(tableDesc, index, s.EndKey, lastPartInclusive)
```

Comment:
Sorry, I know this merged but I was working with this code and was confused for a while. Something non-obvious is going on here - we are not "tightening" the end key. If lastPartInclusive, we are actually "loosening" it to include the value pointing at `EndKey` (that's what `PrefixEnd()` used to do). Granted, we are loosening it as little as possible, but the name of the function is misleading. The comment here should also explain this.

Reply:
Ah, wait, no, we are doing `PrefixEnd` above, line 1147. I don't understand why the old code did `PrefixEnd()` here

Reply:
Ah, no, that's the start key, forget the second comment. The first one still stands.

Reply:
Ah good point. Perhaps `AdjustEndKey` (and `AdjustStartKey`) with said comment(s)? It's less apparent what it does by the name but it's more "correct".

Reply:
Yeah, that works. Though I think `TightenStartKey` is accurate?
Also, I don't understand why the old code did `PrefixEnd` even in the `!lastPartInclusive` case?

Reply:
If I understand your question correctly, the old code short-circuited if the last part is `!inclusive` here [at line 1160](https://github.com/cockroachdb/cockroach/pull/20235/files/048ec01a82dd5e01b9d0602b2cf575db2b68aef0#diff-e7c85a21eb65e3769380c3a0cbe91cc5L1160)

Reply:
Ah! Got it, thanks!